### PR TITLE
Automatically use ReleaseExe when top-level statements are present in a compilation and no options are specified.

### PR DESCRIPTION
**Pull request by 333fred**

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a quick review of this test-only PR that we discussed in the meeting today.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IObjectCreationExpression.cs [hunk 1]
```diff
@@ -799,7 +799,7 @@ public void M1()
         [Fact, WorkItem(1198816, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1198816/")]
         public void ImplicitObjectCreationUnconverted_ArrayIndex()
         {
-            var comp = CreateCompilation("_ = new int[/*<bind>*/new(bad)/*</bind>*/];", options: TestOptions.ReleaseExe);
+            var comp = CreateCompilation("_ = new int[/*<bind>*/new(bad)/*</bind>*/];");
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IObjectCreationExpression.cs:802]: Removed a couple of these from tests I'd recently created to make sure it was working.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 2]
```diff
@@ -82,7 +82,7 @@ class Point(int x, int y);
                 // class Point(int x, int y);
                 Diagnostic(ErrorCode.ERR_UseDefViolation, "int y").WithArguments("y").WithLocation(2, 20)
             );
-            comp = CreateCompilation(src2, parseOptions: TestOptions.Regular8);
+            comp = CreateCompilation(src2, parseOptions: TestOptions.Regular8, options: TestOptions.ReleaseDll);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:85]: >, options: TestOptions.ReleaseDll [](start = 77, length = 33)

Is this change necessary?

File: src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs [hunk 3]
```diff
@@ -293,7 +293,7 @@ public class B
 
             var countedTree = new CountedSyntaxTree(foreignType);
 
-            var compilation = CreateCompilation(new SyntaxTree[] { underlyingTree, countedTree }, skipUsesIsNullable: true);
+            var compilation = CreateCompilation(new SyntaxTree[] { underlyingTree, countedTree }, skipUsesIsNullable: true, options: TestOptions.ReleaseDll);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs:296]: >, options: TestOptions.ReleaseDll [](start = 122, length = 33)

Is this change necessary?

File: src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs [hunk 4]
```diff
@@ -1090,9 +1091,13 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
             bool skipUsesIsNullable,
             MessageID? experimentalFeature)
         {
+            var syntaxTrees = source.GetSyntaxTrees(parseOptions, sourceFileName);
+
             if (options == null)
             {
-                options = TestOptions.ReleaseDll;
+                bool hasTopLevelStatements = syntaxTrees.Any(s => s.GetRoot().DescendantNodes().OfType<GlobalStatementSyntax>().Any());
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:1098]: >DescendantNodes() [](start = 78, length = 17)

Does this look too deep? I think we should check only immediate children of a compilation unit.
See
```
        internal static bool IsTopLevelStatement([NotNullWhen(true)] GlobalStatementSyntax? syntax)
        {
            return syntax?.Parent?.IsKind(SyntaxKind.CompilationUnit) == true;
        }
```

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs:296]: Yes, it changes all the `AccessCounts` to 2 instead of 1.

---
In reply to: [554043500](https://github.com/dotnet/roslyn/pull/50325#discussion_r554043500) [](ancestors = 554043500)

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [756852938](https://github.com/dotnet/roslyn/pull/50325#issuecomment-756852938) [](ancestors = 756852938)

File: src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs:296]: It feels suspicious that changing the output kind would have this effect. I think we need to understand what is going on.

---
In reply to: [554099386](https://github.com/dotnet/roslyn/pull/50325#discussion_r554099386) [](ancestors = 554099386,554043500)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

File: src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs (hunk 3 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/DeclarationTests.cs:296]: Ah, ok. It's because calls to `GetRoot()` are counted, so the new check for "does this syntax have top-level statements" increments the counter. I'll add a remark to `CountedSyntaxTree` about it.

---
In reply to: [554109951](https://github.com/dotnet/roslyn/pull/50325#discussion_r554109951) [](ancestors = 554109951,554099386,554043500)
